tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -194,20 +236,23 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
     val chainProto = tree.superClass.fold[js.Tree] {
       js.Skip()
     } { parentIdent =>
+      val (inheritedCtorDef, inheritedCtorRef) = if (!isJSClass) {
+        (js.Skip(), envField("h", parentIdent.name))
+      } else {
+        val superCtor = genRawJSClassConstructor(
+            linkedClassByName(parentIdent.name))
+        (makeInheritableCtorDef(superCtor), envField("h", className))
+      }
       js.Block(
-        js.Assign(typeVar.prototype,
-            js.New(envField("h", parentIdent.name, parentIdent.originalName), Nil)),
-        genAddToPrototype(className, js.Ident("constructor"), typeVar)
+          inheritedCtorDef,
+          js.Assign(typeVar.prototype, js.New(inheritedCtorRef, Nil)),
+          genAddToPrototype(className, js.Ident("constructor"), typeVar)
       )
     }
 
-    val inheritableCtorDef = {
-      js.Block(
-        js.DocComment("@constructor"),
-        envFieldDef("h", className, js.Function(Nil, js.Skip())),
-        js.Assign(envField("h", className).prototype, typeVar.prototype)
-      )
-    }
+    val inheritableCtorDef =
+      if (isJSClass) js.Skip()
+      else makeInheritableCtorDef(typeVar)
```

Comment:
Hmmm... OK. So you just don't want to do the space optimization here? Since we are at link time, an extending class can just check the type of its superclass and emit an inheritable ctor or not based on that.